java-checks/src/main/java/org/sonar/java/checks/ValueBasedObjectsShouldNotBeSerializedCheck.java

```diff
@@ -0,0 +1,91 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableList;
+import java.util.List;
+import org.sonar.check.Rule;
+import org.sonar.java.checks.helpers.ValueBasedUtils;
+import org.sonar.java.checks.serialization.SerializableContract;
+import org.sonar.java.model.ModifiersUtils;
+import org.sonar.java.resolve.JavaType;
+import org.sonar.java.resolve.ParametrizedTypeJavaType;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.Modifier;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.Tree.Kind;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+@Rule(key = "S3437")
+public class ValueBasedObjectsShouldNotBeSerializedCheck extends IssuableSubscriptionVisitor {
+
+  @Override
+  public List<Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.CLASS);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    ClassTree classTree = (ClassTree)tree;
+    if (isSerializable(classTree) && !SerializableContract.hasSpecialHandlingSerializationMethods(classTree)) {
+      classTree.members().stream()
+        .filter(member -> member.is(Tree.Kind.VARIABLE))
+        .map(member -> (VariableTree) member)
+        .filter(var -> !isStatic(var))
+        .filter(var -> !isTransient(var))
+        .filter(var -> isSerializableAndValueBased(var.type().symbolType()))
+        .forEach(var -> reportIssue(var.simpleName(),
+          "Make this value-based field transient so it is not included in the serialization of this class."));
+    }
+  }
+
+  protected static boolean isSerializableAndValueBased(Type type) {
+    // we check first the ParametrizedTypeJavaType, in order to filter out the non-serializable
+    // generic value-based class Optional<T>
+    JavaType javaType = (JavaType)type;
+    if (javaType.isParameterized()) {
+      ParametrizedTypeJavaType parameterizedType = (ParametrizedTypeJavaType)javaType;
+      return isSubtypeOfCollectionApi(parameterizedType) &&
+             parameterizedType.typeParameters().stream()
+               .anyMatch(t -> isSerializableAndValueBased(parameterizedType.substitution(t)));
+    } else {
+      return ValueBasedUtils.isValueBased(javaType);
+    }
+  }
+
+  private static boolean isSerializable(ClassTree classTree) {
+    return classTree.symbol().type().isSubtypeOf("java.io.Serializable");
+  }
+
+  private static boolean isStatic(VariableTree variable) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this unused private "isStatic" method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUnusedPrivateMethod)